subiquity/server/controllers/keyboard.py

```diff
@@ -203,16 +204,26 @@ async def set_keyboard(self):
             ['setupcon', '--save', '--force', '--keyboard-only'],
             [resource_path('bin/subiquity-loadkeys')],
             ]
-        if shutil.which('setxkbmap'):
-            setxkbmap = ['setxkbmap', '-layout', self.model.setting.layout]
-            if self.model.setting.variant:
-                setxkbmap.extend(['-variant', self.model.setting.variant])
-            cmds.append(setxkbmap)
         if self.opts.dry_run:
             scale = os.environ.get('SUBIQUITY_REPLAY_TIMESCALE', "1")
             cmds = [['sleep', str(1/float(scale))]]
         for cmd in cmds:
             await arun_command(cmd)
+        await self.run_setxkbmap()
+
+    async def run_setxkbmap(self):
+        if not self.opts.dry_run and shutil.which('setxkbmap'):
+            try:
+                # seteuid() because Popen(user=) is new in 3.9 (UDI has 3.8)
+                os.seteuid(pwd.getpwnam('ubuntu').pw_uid)
```

Comment:
`run_setxkbmap` is an asynchronous function so there is a chance that a context switch occurs before the EUID gets reset by the `finally` block. I think this can lead to some hard to reproduce bugs.
Would you mind trying the following construct instead?
```python
setxkbmap = ["runuser", "-u", "ubuntu", "--", "setxkbmap", "-layout", self.model.setting.layout]
```
`runuser` is provided by `util-linux` ; and I believe it should mimic the `user` parameter of `subprocess.Popen` quite well.

Reply:
Nice idea. Manually executing `runuser` as root in a clean environment works, but I'm having problems running `runuser` from Subiquity:
```
runuser: cannot open session: Module is unknown
```
Seems to be an error from `pam_open_session()`. :thinking:

Reply:
I suppose it's essentially the same problem why `sudo -u ubuntu` didn't work either but I had to set the effective uid.

Reply:
pam_open_session(3):
> It should be noted that the effective uid, geteuid. of the application should be of sufficient privilege to perform such tasks as creating or mounting the user's home directory for example.

Reply:
yikes!
I guess this would essentially work but this feels like a stretch ...
```python
command = """\
import subprocess
import os
import pwd
import sys

os.seteuid(pwd.getpwnam("ubuntu").pw_uid)
subprocess.run(("setxkbmap", "-layout", sys.argv[1]))
"""

setxkbmap = ["python3", "-c", command, self.model.setting.layout]
```

Reply:
I'll take a stab at upgrading UDI from core20 to core22. Then we'd have Python 3.10 available...

Reply:
Subiquity on server runs on core20. We haven't been doing testing on python 3.10 AFAIK so I would not recommend it.
Are you sure subprocess.Popen(user="ubuntu") would not fail with a similar error to "cannot open session: Module is unknown" or something?

Reply:
`subprocess.Popen(user="ubuntu")` calls `setreuid()` under the hood.

Reply:
What if we'd make it sync? setxkbmap takes no time to execute.

Reply:
This seems to work
```py
    async def run_setxkbmap(self):
        if not self.opts.dry_run and shutil.which('setxkbmap'):
            command = """\
import subprocess
import os
import pwd
import sys

os.seteuid(pwd.getpwnam("ubuntu").pw_uid)
subprocess.run(["setxkbmap", "-layout", sys.argv[1], "-variant", sys.argv[2]])
"""
            variant = self.model.setting.variant
            setxkbmap = ["python3", "-c", command, self.model.setting.layout,
                         variant if not None else ""]
            await arun_command(setxkbmap)
```

Reply:
Coming back to this... I just realized the desktop installer has some logic to run a specific Python interpreter a) depending on whether it's snapped and b) to avoid conflicts with the Python interpreter in the Flutter snap. I'm not sure if it matters here, but this code would run a different Python interpreter.

Reply:
It looks like `sys.executable` gives the correct Python interpreter. Should we proceed with this?

Reply:
Yes please, I am more comfortable with this approach.

`sys.executable` - or alternatively `os.environ["PYTHON"]` - is the right way to do it :+1: 

Is setxkbmap okay with receiving an empty string as the value for `-variant` or should we pass `-variant` only if the value is provided?